File: Week-Wise-Assignments/Assignment 7/Systematic Solutions/Python Programs/3_insertion_sort.py
def insertion_sort(arr, show_steps=False):
    """
    Algorithm for Insertion Sort:
    1. Start with second element (assume first is sorted)
    2. Compare current element with elements in sorted portion
    3. Shift larger elements to the right
    4. Insert current element at correct position
    5. Repeat for all elements
    """
    n = len(arr)
    
    if show_steps:
        print(f"\n=== Insertion Sort Algorithm ===")
        print(f"Array size: {n}")
        print(f"Initial array: {arr}")
        print("Algorithm: Insert each element at its correct position in sorted portion\n")
    
    total_comparisons = 0
    total_shifts = 0
    
    for i in range(1, n):
        key = arr[i]
        j = i - 1
        pass_comparisons = 0
        pass_shifts = 0
        
        if show_steps:
            print(f"Pass {i}: Inserting arr[{i}] = {key} into sorted portion [0, {i-1}]")
            print(f"  Key to insert: {key}")
            print(f"  Sorted portion: {arr[:i]}")
        
        # Move elements that are greater than key one position ahead
        while j >= 0:
            pass_comparisons += 1
            total_comparisons += 1
            
            if show_steps:
                print(f"  Comparing key {key} with arr[{j}] = {arr[j]}: ", end="")
            
            if arr[j] > key:
                if show_steps:
                    print(f"arr[{j}] > key, shifting right")
                
                arr[j + 1] = arr[j]
                j -= 1
                pass_shifts += 1
                total_shifts += 1
                
                if show_steps:
                    print(f"    After shift: {arr}")
            else:
                if show_steps:
                    print(f"arr[{j}] <= key, found insertion position")
                break
        
        # Insert key at correct position
        arr[j + 1] = key
        
        if show_steps:
            print(f"  Inserted {key} at position {j + 1}")
            print(f"  After pass {i}: {arr}")
            print(f"  Sorted portion: [0, {i}], Unsorted portion: [{i + 1}, {n - 1}]")
            print(f"  Pass statistics: {pass_comparisons} comparisons, {pass_shifts} shifts")
            print()
    
    if show_steps:
        print("Insertion Sort completed!")
        print(f"Total statistics: {total_comparisons} comparisons, {total_shifts} shifts")
        print(f"Final sorted array: {arr}")
    
    return arr, total_comparisons, total_shifts

def binary_insertion_sort(arr):
    """Binary insertion sort - uses binary search to find insertion position"""
    n = len(arr)
    print(f"\n=== Binary Insertion Sort ===")
    print(f"Initial array: {arr}")
    
    total_comparisons = 0
    total_shifts = 0
    
    for i in range(1, n):
        key = arr[i]
        left, right = 0, i - 1
        binary_comparisons = 0
        
        print(f"Pass {i}: Finding position for {key} using binary search")
        
        # Binary search to find insertion position
        while left <= right:
            mid = left + (right - left) // 2
            binary_comparisons += 1
            total_comparisons += 1
            
            print(f"  Binary search: left={left}, right={right}, mid={mid}, arr[mid]={arr[mid]}")
            
            if arr[mid] > key:
                right = mid - 1
            else:
                left = mid + 1
        
        # Shift elements and insert
        insert_pos = left
        print(f"  Insertion position found: {insert_pos}")
        
        shifts = i - insert_pos
        for j in range(i - 1, insert_pos - 1, -1):
            arr[j + 1] = arr[j]
            total_shifts += 1
        
        arr[insert_pos] = key
        
        print(f"  After inserting {key}: {arr}")
        print(f"  Binary search comparisons: {binary_comparisons}, Shifts: {shifts}")
        print()
    
    print("Binary Insertion Sort completed!")
    print(f"Total statistics: {total_comparisons} comparisons, {total_shifts} shifts")
    
    return arr

def performance_analysis():
    """Analyze performance for different test cases"""
    print(f"\n=== Performance Analysis ===")
    
    sizes = [10, 50, 100, 500]
    
    print("Test Case Analysis:")
    print("Array Size | Best Case | Average Case | Worst Case")
    print("-----------|-----------|--------------|------------")
    
    for size in sizes:
        # Best case: already sorted
        best_comparisons = size - 1
        
        # Average case approximation
        avg_comparisons = sum(i // 2 for i in range(1, size))
        
        # Worst case: reverse sorted
        worst_comparisons = sum(range(1, size))
        
        print(f"{size:10d} | {best_comparisons:9d} | {avg_comparisons:12d} | {worst_comparisons:10d}")

def generate_test_array(size, array_type):
    """Generate different types of test arrays"""
    import random
    
    if array_type == 1:  # Random array
        return [random.randint(1, 100) for _ in range(size)]
    elif array_type == 2:  # Nearly sorted array
        arr = list(range(1, size + 1))
        # Swap a few elements
        if size > 4:
            arr[1], arr[3] = arr[3], arr[1]
        return arr
    elif array_type == 3:  # Reverse sorted array
        return list(range(size, 0, -1))
    elif array_type == 4:  # Array with duplicates
        return [random.randint(1, 5) for _ in range(size)]
    else:
        return [random.randint(1, 100) for _ in range(size)]

def is_sorted(arr):
    """Check if array is sorted"""
    return all(arr[i] <= arr[i + 1] for i in range(len(arr) - 1))

def main():
    """Main function with interactive menu"""
    print("=== Insertion Sort Implementation ===")
    print("Insertion Sort: Builds sorted array one element at a time")
    print("Time Complexity: O(n) best case, O(n²) average and worst case")
    print("Space Complexity: O(1) - In-place sorting")
    print("Stability: Stable (maintains relative order of equal elements)\n")
    
    while True:
        print("\n--- Menu ---")
        print("1. Sort custom array (step-by-step)")
        print("2. Binary insertion sort")
        print("3. Sort test arrays (different types)")
        print("4. Performance analysis")
        print("5. Algorithm explanation")
        print("6. Exit")
        
        try:
            choice = int(input("Enter your choice: "))
            
            if choice == 1:
                size = int(input("Enter array size: "))
                
                if size <= 0 or size > 100:
                    print("Invalid size! Please enter size between 1 and 100.")
                    continue
                
                print(f"Enter {size} elements separated by spaces: ")
                arr = list(map(int, input().split()))
                
                if len(arr) != size:
                    print("Number of elements doesn't match the specified size!")
                    continue
                
                sorted_arr, comparisons, shifts = insertion_sort(arr, show_steps=True)
                
                print(f"\nVerification: Array is {'correctly sorted' if is_sorted(sorted_arr) else 'NOT sorted'}")
                
            elif choice == 2:
                size = int(input("Enter array size: "))
                
                if size <= 0 or size > 50:
                    print("Invalid size! Please enter size between 1 and 50.")
                    continue
                
                print(f"Enter {size} elements separated by spaces: ")
                arr = list(map(int, input().split()))
                
                if len(arr) != size:
                    print("Number of elements doesn't match the specified size!")
                    continue
                
                binary_insertion_sort(arr)
                
            elif choice == 3:
                size = int(input("Enter array size: "))
                
                if size <= 0 or size > 100:
                    print("Invalid size! Please enter size between 1 and 100.")
                    continue
                
                print("\nSelect array type:")
                print("1. Random array")
                print("2. Nearly sorted array")
                print("3. Reverse sorted array")
                print("4. Array with duplicates")
                
                array_type = int(input("Enter choice: "))
                
                if array_type < 1 or array_type > 4:
                    print("Invalid choice!")
                    continue
                
                arr = generate_test_array(size, array_type)
                
                type_names = ["", "Random", "Nearly Sorted", "Reverse Sorted", "With Duplicates"]
                print(f"\nGenerated {type_names[array_type]} Array:")
                
                insertion_sort(arr, show_steps=True)
                
            elif choice == 4:
                performance_analysis()
                
            elif choice == 5:
                print(f"\n=== Insertion Sort Algorithm Explanation ===")
                print("How it works:")
                print("1. Start with the second element (consider first element as sorted)")
                print("2. Compare current element with elements in the sorted portion")
                print("3. Shift all larger elements one position to the right")
                print("4. Insert current element at the correct position")
                print("5. Repeat for all remaining elements\n")
                
                print("Characteristics:")
                print("• Simple and intuitive algorithm")
                print("• Stable sorting algorithm")
                print("• In-place sorting (O(1) extra space)")
                print("• Adaptive (efficient for nearly sorted arrays)")
                print("• Online (can sort array as it receives data)\n")
                
                print("Time Complexity:")
                print("• Best Case: O(n) - when array is already sorted")
                print("• Average Case: O(n²)")
                print("• Worst Case: O(n²) - when array is reverse sorted\n")
                
                print("Variants:")
                print("• Binary Insertion Sort: Uses binary search to find insertion position")
                print("  - Reduces comparisons but same number of shifts")
                print("  - Time complexity remains O(n²) due to shifting\n")
                
                print("When to use:")
                print("• Small datasets (typically n < 50)")
                print("• Nearly sorted arrays")
                print("• Online algorithms (sorting data as it arrives)")
                print("• As a subroutine in hybrid algorithms (like Quicksort)")
                
            elif choice == 6:
                print("Exiting program...")
                break
                
            else:
                print("Invalid choice! Please try again.")
                
        except ValueError:
            print("Invalid input! Please enter a valid number.")
        except Exception as e:
            print(f"An error occurred: {e}")

if __name__ == "__main__":
    main()

File: Week-Wise-Assignments/Assignment 7/Systematic Solutions/C Programs/3_insertion_sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

// Function to display array
void displayArray(int arr[], int n, const char* message) {
    printf("%s: ", message);
    for (int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

// Algorithm for Insertion Sort:
// 1. Start with second element (assume first is sorted)
// 2. Compare current element with elements in sorted portion
// 3. Shift larger elements to the right
// 4. Insert current element at correct position
// 5. Repeat for all elements
void insertionSort(int arr[], int n) {
    printf("\n=== Insertion Sort Algorithm ===\n");
    printf("Array size: %d\n", n);
    printf("Algorithm: Insert each element at its correct position in sorted portion\n\n");
    
    displayArray(arr, n, "Initial array");
    
    int totalComparisons = 0;
    int totalShifts = 0;
    
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int j = i - 1;
        int passComparisons = 0;
        int passShifts = 0;
        
        printf("Pass %d: Inserting arr[%d] = %d into sorted portion [0, %d]\n", 
               i, i, key, i - 1);
        printf("  Key to insert: %d\n", key);
        printf("  Sorted portion: ");
        for (int k = 0; k < i; k++) {
            printf("%d ", arr[k]);
        }
        printf("\n");
        
        // Move elements that are greater than key one position ahead
        while (j >= 0) {
            passComparisons++;
            totalComparisons++;
            
            printf("  Comparing key %d with arr[%d] = %d: ", key, j, arr[j]);
            
            if (arr[j] > key) {
                printf("arr[%d] > key, shifting right\n", j);
                arr[j + 1] = arr[j];
                j--;
                passShifts++;
                totalShifts++;
                
                // Show array state after shift
                printf("    After shift: ");
                for (int k = 0; k < n; k++) {
                    printf("%d ", arr[k]);
                }
                printf("\n");
            } else {
                printf("arr[%d] <= key, found insertion position\n", j);
                break;
            }
        }
        
        // Insert key at correct position
        arr[j + 1] = key;
        printf("  Inserted %d at position %d\n", key, j + 1);
        
        printf("  After pass %d: ", i);
        displayArray(arr, n, "");
        printf("  Sorted portion: [0, %d], Unsorted portion: [%d, %d]\n", 
               i, i + 1, n - 1);
        printf("  Pass statistics: %d comparisons, %d shifts\n", 
               passComparisons, passShifts);
        printf("\n");
    }
    
    printf("Insertion Sort completed!\n");
    printf("Total statistics: %d comparisons, %d shifts\n", 
           totalComparisons, totalShifts);
    displayArray(arr, n, "Final sorted array");
}

// Binary insertion sort - uses binary search to find insertion position
void binaryInsertionSort(int arr[], int n) {
    printf("\n=== Binary Insertion Sort ===\n");
    displayArray(arr, n, "Initial array");
    
    int totalComparisons = 0;
    int totalShifts = 0;
    
    for (int i = 1; i < n; i++) {
        int key = arr[i];
        int left = 0, right = i - 1;
        int binaryComparisons = 0;
        
        printf("Pass %d: Finding position for %d using binary search\n", i, key);
        
        // Binary search to find insertion position
        while (left <= right) {
            int mid = left + (right - left) / 2;
            binaryComparisons++;
            totalComparisons++;
            
            printf("  Binary search: left=%d, right=%d, mid=%d, arr[mid]=%d\n", 
                   left, right, mid, arr[mid]);
            
            if (arr[mid] > key) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }
        
        // Shift elements and insert
        int insertPos = left;
        printf("  Insertion position found: %d\n", insertPos);
        
        for (int j = i - 1; j >= insertPos; j--) {
            arr[j + 1] = arr[j];
            totalShifts++;
        }
        arr[insertPos] = key;
        
        printf("  After inserting %d: ", key);
        displayArray(arr, n, "");
        printf("  Binary search comparisons: %d, Shifts: %d\n", 
               binaryComparisons, i - insertPos);
        printf("\n");
    }
    
    printf("Binary Insertion Sort completed!\n");
    printf("Total statistics: %d comparisons, %d shifts\n", 
           totalComparisons, totalShifts);
}

// Performance analysis function
void performanceAnalysis() {
    printf("\n=== Performance Analysis ===\n");
    
    int sizes[] = {10, 50, 100, 500};
    int numSizes = sizeof(sizes) / sizeof(sizes[0]);
    
    printf("Test Case Analysis:\n");
    printf("Array Size | Best Case | Average Case | Worst Case\n");
    printf("-----------|-----------|--------------|------------\n");
    
    for (int i = 0; i < numSizes; i++) {
        int n = sizes[i];
        
        // Best case: already sorted
        int* bestCase = (int*)malloc(n * sizeof(int));
        for (int j = 0; j < n; j++) {
            bestCase[j] = j + 1;
        }
        
        // Average case: random
        int* avgCase = (int*)malloc(n * sizeof(int));
        srand(time(NULL) + i);
        for (int j = 0; j < n; j++) {
            avgCase[j] = rand() % 1000;
        }
        
        // Worst case: reverse sorted
        int* worstCase = (int*)malloc(n * sizeof(int));
        for (int j = 0; j < n; j++) {
            worstCase[j] = n - j;
        }
        
        // Count comparisons for each case
        int bestComparisons = n - 1; // Best case always n-1 comparisons
        
        // Average case simulation
        int avgComparisons = 0;
        for (int pass = 1; pass < n; pass++) {
            avgComparisons += pass / 2; // Average case approximation
        }
        
        // Worst case
        int worstComparisons = 0;
        for (int pass = 1; pass < n; pass++) {
            worstComparisons += pass;
        }
        
        printf("%10d | %9d | %12d | %10d\n", 
               n, bestComparisons, avgComparisons, worstComparisons);
        
        free(bestCase);
        free(avgCase);
        free(worstCase);
    }
}

// Function to generate different types of arrays
void generateTestArray(int arr[], int n, int type) {
    switch (type) {
        case 1: // Random array
            srand(time(NULL));
            for (int i = 0; i < n; i++) {
                arr[i] = rand() % 100 + 1;
            }
            break;
            
        case 2: // Nearly sorted array
            for (int i = 0; i < n; i++) {
                arr[i] = i + 1;
            }
            // Swap a few elements
            if (n > 4) {
                int temp = arr[1];
                arr[1] = arr[3];
                arr[3] = temp;
            }
            break;
            
        case 3: // Reverse sorted array
            for (int i = 0; i < n; i++) {
                arr[i] = n - i;
            }
            break;
            
        case 4: // Array with duplicates
            srand(time(NULL));
            for (int i = 0; i < n; i++) {
                arr[i] = rand() % 5 + 1; // Values 1-5
            }
            break;
    }
}

// Function to check if array is sorted
int isSorted(int arr[], int n) {
    for (int i = 0; i < n - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return 0;
        }
    }
    return 1;
}

int main() {
    int choice, n;
    
    printf("=== Insertion Sort Implementation ===\n");
    printf("Insertion Sort: Builds sorted array one element at a time\n");
    printf("Time Complexity: O(n) best case, O(n²) average and worst case\n");
    printf("Space Complexity: O(1) - In-place sorting\n");
    printf("Stability: Stable (maintains relative order of equal elements)\n\n");
    
    while (1) {
        printf("\n--- Menu ---\n");
        printf("1. Sort custom array (step-by-step)\n");
        printf("2. Binary insertion sort\n");
        printf("3. Sort test arrays (different types)\n");
        printf("4. Performance analysis\n");
        printf("5. Algorithm explanation\n");
        printf("6. Exit\n");
        printf("Enter your choice: ");
        scanf("%d", &choice);
        
        switch (choice) {
            case 1: {
                printf("Enter array size: ");
                scanf("%d", &n);
                
                if (n <= 0 || n > 100) {
                    printf("Invalid size! Please enter size between 1 and 100.\n");
                    break;
                }
                
                int* arr = (int*)malloc(n * sizeof(int));
                
                printf("Enter %d elements: ", n);
                for (int i = 0; i < n; i++) {
                    scanf("%d", &arr[i]);
                }
                
                insertionSort(arr, n);
                
                printf("\nVerification: Array is %s\n", 
                       isSorted(arr, n) ? "correctly sorted" : "NOT sorted");
                
                free(arr);
                break;
            }
            
            case 2: {
                printf("Enter array size: ");
                scanf("%d", &n);
                
                if (n <= 0 || n > 50) {
                    printf("Invalid size! Please enter size between 1 and 50.\n");
                    break;
                }
                
                int* arr = (int*)malloc(n * sizeof(int));
                
                printf("Enter %d elements: ", n);
                for (int i = 0; i < n; i++) {
                    scanf("%d", &arr[i]);
                }
                
                binaryInsertionSort(arr, n);
                free(arr);
                break;
            }
            
            case 3: {
                printf("Enter array size: ");
                scanf("%d", &n);
                
                if (n <= 0 || n > 100) {
                    printf("Invalid size! Please enter size between 1 and 100.\n");
                    break;
                }
                
                printf("\nSelect array type:\n");
                printf("1. Random array\n");
                printf("2. Nearly sorted array\n");
                printf("3. Reverse sorted array\n");
                printf("4. Array with duplicates\n");
                printf("Enter choice: ");
                
                int type;
                scanf("%d", &type);
                
                if (type < 1 || type > 4) {
                    printf("Invalid choice!\n");
                    break;
                }
                
                int* arr = (int*)malloc(n * sizeof(int));
                generateTestArray(arr, n, type);
                
                const char* typeNames[] = {"", "Random", "Nearly Sorted", "Reverse Sorted", "With Duplicates"};
                printf("\nGenerated %s Array:\n", typeNames[type]);
                
                insertionSort(arr, n);
                free(arr);
                break;
            }
            
            case 4:
                performanceAnalysis();
                break;
                
            case 5:
                printf("\n=== Insertion Sort Algorithm Explanation ===\n");
                printf("How it works:\n");
                printf("1. Start with the second element (consider first element as sorted)\n");
                printf("2. Compare current element with elements in the sorted portion\n");
                printf("3. Shift all larger elements one position to the right\n");
                printf("4. Insert current element at the correct position\n");
                printf("5. Repeat for all remaining elements\n\n");
                
                printf("Characteristics:\n");
                printf("• Simple and intuitive algorithm\n");
                printf("• Stable sorting algorithm\n");
                printf("• In-place sorting (O(1) extra space)\n");
                printf("• Adaptive (efficient for nearly sorted arrays)\n");
                printf("• Online (can sort array as it receives data)\n\n");
                
                printf("Time Complexity:\n");
                printf("• Best Case: O(n) - when array is already sorted\n");
                printf("• Average Case: O(n²)\n");
                printf("• Worst Case: O(n²) - when array is reverse sorted\n\n");
                
                printf("Variants:\n");
                printf("• Binary Insertion Sort: Uses binary search to find insertion position\n");
                printf("  - Reduces comparisons but same number of shifts\n");
                printf("  - Time complexity remains O(n²) due to shifting\n\n");
                
                printf("When to use:\n");
                printf("• Small datasets (typically n < 50)\n");
                printf("• Nearly sorted arrays\n");
                printf("• Online algorithms (sorting data as it arrives)\n");
                printf("• As a subroutine in hybrid algorithms (like Quicksort)\n");
                break;
                
            case 6:
                printf("Exiting program...\n");
                exit(0);
                
            default:
                printf("Invalid choice! Please try again.\n");
        }
    }
    
    return 0;
}